File: artifacts/java/src/main/java/io/pgenie/artifacts/myspace/musiccatalogue/statements/SelectAlbumFields.java
package io.pgenie.artifacts.myspace.musiccatalogue.statements;

import io.pgenie.artifacts.myspace.musiccatalogue.Statement;
import io.pgenie.artifacts.myspace.musiccatalogue.codecs.Jdbc;
import io.pgenie.artifacts.myspace.musiccatalogue.types.*;

import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Date;
import java.time.*;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

/**
 * Type-safe binding for the {@code select_album_fields} query.
 *
 * <h2>SQL Template</h2>
 *
 * <pre>{@code
 * -- Demonstrates static query equivalent of dynamic field selection.
 * -- Boolean flags control which fields are included in the result,
 * -- returning NULL for fields the caller opts out of.
 * SELECT
 *   album.id,
 *   CASE WHEN $include_name      THEN album.name      END AS name,
 *   CASE WHEN $include_released  THEN album.released  END AS released,
 *   CASE WHEN $include_format    THEN album.format    END AS format,
 *   CASE WHEN $include_recording THEN album.recording END AS recording,
 *   CASE WHEN $include_tracks    THEN album.tracks    END AS tracks,
 *   CASE WHEN $include_disc      THEN album.disc      END AS disc
 * FROM album
 * WHERE album.id = $id
 * }</pre>
 *
 * <h2>Source Path</h2> {@code ./queries/select_album_fields.sql}
 *
 * <p>
 * Generated from SQL queries using the
 * <a href="https://pgenie.io">pGenie</a> code generator.
 */
public record SelectAlbumFields(
        /**
         * Maps to {@code $include_name} in the template.
         */
        boolean includeName,
        /**
         * Maps to {@code $include_released} in the template.
         */
        boolean includeReleased,
        /**
         * Maps to {@code $include_format} in the template.
         */
        boolean includeFormat,
        /**
         * Maps to {@code $include_recording} in the template.
         */
        boolean includeRecording,
        /**
         * Maps to {@code $include_tracks} in the template.
         */
        boolean includeTracks,
        /**
         * Maps to {@code $include_disc} in the template.
         */
        boolean includeDisc,
        /**
         * Maps to {@code $id} in the template.
         */
        long id)
        implements Statement<SelectAlbumFields.Output> {

    // -------------------------------------------------------------------------
    // Result type
    // -------------------------------------------------------------------------
    /**
     * Result of the statement parameterised by {@link SelectAlbumFields}.
     */
    public static final class Output extends ArrayList<OutputRow> {

        Output() {
        }
    }

    /**
     * Row of {@link Output}.
     */
    public record OutputRow(
            /**
             * Maps to the {@code id} result-set column.
             */
            long id,
            /**
             * Maps to the {@code name} result-set column. Nullable.
             */
            Optional<String> name,
            /**
             * Maps to the {@code released} result-set column. Nullable.
             */
            Optional<LocalDate> released,
            /**
             * Maps to the {@code format} result-set column. Nullable.
             */
            Optional<AlbumFormat> format,
            /**
             * Maps to the {@code recording} result-set column. Nullable.
             */
            Optional<RecordingInfo> recording,
            /**
             * Maps to the {@code tracks} result-set column. Nullable.
             */
            Optional<List<TrackInfo>> tracks,
            /**
             * Maps to the {@code disc} result-set column. Nullable.
             */
            Optional<DiscInfo> disc) {

    }

    // -------------------------------------------------------------------------
    // Statement implementation
    // -------------------------------------------------------------------------
    @Override
    public String sql() {
        return """
               -- Demonstrates static query equivalent of dynamic field selection.
               -- Boolean flags control which fields are included in the result,
               -- returning NULL for fields the caller opts out of.
               SELECT
                 album.id,
                 CASE WHEN ?      THEN album.name      END AS name,
                 CASE WHEN ?  THEN album.released  END AS released,
                 CASE WHEN ?    THEN album.format    END AS format,
                 CASE WHEN ? THEN album.recording END AS recording,
                 CASE WHEN ?    THEN album.tracks    END AS tracks,
                 CASE WHEN ?      THEN album.disc      END AS disc
               FROM album
               WHERE album.id = ?
               """;
    }

    @Override
    public void bindParams(PreparedStatement ps) throws SQLException {
        ps.setBoolean(1, this.includeName());
        ps.setBoolean(2, this.includeReleased());
        ps.setBoolean(3, this.includeFormat());
        ps.setBoolean(4, this.includeRecording());
        ps.setBoolean(5, this.includeTracks());
        ps.setBoolean(6, this.includeDisc());
        ps.setLong(7, this.id());
    }

    @Override
    public boolean returnsRows() {
        return true;
    }

    @Override
    public Output decodeResultSet(ResultSet rs) throws SQLException {
        Output output = new Output();
        while (rs.next()) {
            try {
                long id = rs.getLong(1);
                Optional<String> name = Optional.ofNullable(rs.getString(2));
                Date releasedSql = rs.getDate(3);
                Optional<LocalDate> released = Optional.ofNullable(releasedSql != null ? releasedSql.toLocalDate() : null);
                String formatStr = rs.getString(4);
                Optional<AlbumFormat> format = Optional.ofNullable(formatStr != null ? AlbumFormat.CODEC.decodeInTextFromString(formatStr) : null);
                String recordingStr = rs.getString(5);
                Optional<RecordingInfo> recording = Optional.ofNullable(recordingStr != null ? RecordingInfo.CODEC.decodeInTextFromString(recordingStr) : null);
                String tracksStr = rs.getString(6);
                Optional<List<TrackInfo>> tracks = Optional.ofNullable(tracksStr != null ? TrackInfo.CODEC.inDim().decodeInTextFromString(tracksStr) : null);
                String discStr = rs.getString(7);
                Optional<DiscInfo> disc = Optional.ofNullable(discStr != null ? DiscInfo.CODEC.decodeInTextFromString(discStr) : null);
                output.add(new OutputRow(id, name, released, format, recording, tracks, disc));
            } catch (io.codemine.java.postgresql.codecs.Codec.DecodingException e) {
                throw new IllegalStateException(e);
            }
        }
        return output;
    }

    @Override
    public SelectAlbumFields.Output decodeAffectedRows(long affectedRows) {
        throw new UnsupportedOperationException();
    }
}

File: artifacts/rust/src/statements/select_album_fields.rs
use tokio_postgres::types::Type;

/// Parameters for the `select_album_fields` query.
///
/// # SQL Template
///
/// ```sql
/// -- Demonstrates static query equivalent of dynamic field selection.
/// -- Boolean flags control which fields are included in the result,
/// -- returning NULL for fields the caller opts out of.
/// SELECT
///   album.id,
///   CASE WHEN $include_name      THEN album.name      END AS name,
///   CASE WHEN $include_released  THEN album.released  END AS released,
///   CASE WHEN $include_format    THEN album.format    END AS format,
///   CASE WHEN $include_recording THEN album.recording END AS recording,
///   CASE WHEN $include_tracks    THEN album.tracks    END AS tracks,
///   CASE WHEN $include_disc      THEN album.disc      END AS disc
/// FROM album
/// WHERE album.id = $id
/// ```
///
/// # Source Path
///
/// `./queries/select_album_fields.sql`
#[derive(Debug, Clone, PartialEq, Default)]
pub struct Input {
    /// Maps to `$include_name` in the template.
    pub include_name: bool,
    /// Maps to `$include_released` in the template.
    pub include_released: bool,
    /// Maps to `$include_format` in the template.
    pub include_format: bool,
    /// Maps to `$include_recording` in the template.
    pub include_recording: bool,
    /// Maps to `$include_tracks` in the template.
    pub include_tracks: bool,
    /// Maps to `$include_disc` in the template.
    pub include_disc: bool,
    /// Maps to `$id` in the template.
    pub id: i64,
}

/// Result of the statement parameterised by [`Input`].
pub type Output = Vec<OutputRow>;

/// Row of [`Output`].
#[derive(Debug, Clone, PartialEq)]
pub struct OutputRow {
    /// Maps to the `id` result set column.
    pub id: i64,
    /// Maps to the `name` result set column.
    pub name: Option<String>,
    /// Maps to the `released` result set column.
    pub released: Option<chrono::NaiveDate>,
    /// Maps to the `format` result set column.
    pub format: Option<crate::types::AlbumFormat>,
    /// Maps to the `recording` result set column.
    pub recording: Option<crate::types::RecordingInfo>,
    /// Maps to the `tracks` result set column.
    pub tracks: Option<Vec<crate::types::TrackInfo>>,
    /// Maps to the `disc` result set column.
    pub disc: Option<crate::types::DiscInfo>,
}


impl crate::mapping::Statement for Input {
    type Result = Output;

    const RETURNS_ROWS: bool = true;

    const SQL: &str = "-- Demonstrates static query equivalent of dynamic field selection.\n\
                       -- Boolean flags control which fields are included in the result,\n\
                       -- returning NULL for fields the caller opts out of.\n\
                       SELECT\n\
                         album.id,\n\
                         CASE WHEN $1      THEN album.name      END AS name,\n\
                         CASE WHEN $2  THEN album.released  END AS released,\n\
                         CASE WHEN $3    THEN album.format    END AS format,\n\
                         CASE WHEN $4 THEN album.recording END AS recording,\n\
                         CASE WHEN $5    THEN album.tracks    END AS tracks,\n\
                         CASE WHEN $6      THEN album.disc      END AS disc\n\
                       FROM album\n\
                       WHERE album.id = $7";

    const PARAM_TYPES: &'static [tokio_postgres::types::Type] = &[Type::BOOL, Type::BOOL, Type::BOOL, Type::BOOL, Type::BOOL, Type::BOOL, Type::INT8];

    #[allow(refining_impl_trait)]
    fn encode_params(
        &self,
    ) -> [&(dyn tokio_postgres::types::ToSql + Sync); Self::PARAM_TYPES.len()] {
        [&self.include_name, &self.include_released, &self.include_format, &self.include_recording, &self.include_tracks, &self.include_disc, &self.id]
    }

    fn decode_result(
        rows: Vec<tokio_postgres::Row>,
        _affected_rows: u64,
    ) -> Result<Self::Result, crate::mapping::DecodingError> {
        rows.into_iter()
            .enumerate()
            .map(|(row_index, row)| {
                Ok(OutputRow {
                    id: crate::mapping::decode_cell(&row, row_index, 0)?,
                    name: crate::mapping::decode_cell(&row, row_index, 1)?,
                    released: crate::mapping::decode_cell(&row, row_index, 2)?,
                    format: crate::mapping::decode_cell(&row, row_index, 3)?,
                    recording: crate::mapping::decode_cell(&row, row_index, 4)?,
                    tracks: crate::mapping::decode_cell(&row, row_index, 5)?,
                    disc: crate::mapping::decode_cell(&row, row_index, 6)?,
                })
            })
            .collect()
    }
}
